### ISSUE TITLE
GSoC: Test Function Suite

### ISSUE THREAD

**Issue by pkofod** [JuliaOpt/Optim.jl#223]
I figured I should update details in separate issues instead of the one big issue #200 .

I am planning to start off by testing on CUTEst unconstrained functions, through the CUTEst.jl package. I installed it by following http://abelsiqueira.github.io/blog/installing-cutest-and-cutest.jl/ . The develop branch is necessary, as this has a `grad!` function defined already. The master branch also segfaulted on me quite often.

I had originally chosen the 216 unconstrained problems in CUTEst, but one seemed to have constraints, and CUTEst.jl had problems loading quite a few (I need to file an issue upstream), so now there are closer to 160. 

An interesting feature of the test suite is how broadly we're testing in terms of dimensionality of x (call it N). It seems that a lot of functions are 2- or 3- dimensional, but also values of around 3000 and 5000 seem popular. 11 functions have N = 10000, and 2 have N = 20000. Some of the original formulations allow for a user defined N (like in the Rosenbrock case), but I'm not sure of CUTEst allows the user to specify this.
![nvar_hist](https://cloud.githubusercontent.com/assets/8431156/16060758/aaf1b0f0-3289-11e6-8c72-5ec9dbfcc9b8.png)

Anyway, this is a much broader test bed than we already have, and I'll try to post some initial benchmarks as soon as possible.

**Comment by mlubin** [comment]
CC @abelsiqueira @dpo

**Comment by dpo** [comment]
Note that CUTEst.jl is now available from https://github.com/JuliaOptimizers/CUTEst.jl in more useable form. A lot of the CUTEst problems are also available in AMPL format from https://github.com/mpf/Optimization-Test-Problems and you can read them (once decoded) using https://github.com/JuliaOptimizers/AmplNLReader.jl. If you want to change the size of an AMPL model, you must edit the `mod` file and decode it again. If you want to see what sizes a CUTEst model comes in, you can do, e.g.,
```
sifdecoder -show LUBRIFC
```
and subsequently decode with
```
sifdecoder -param NN=50 LUBRIFC
```
CUTEst.jl doesn't currently support passing extra options to `sifdecoder` (I believe) but it would be easy to add that feature. You can select all unconstrained problems using the `select` tool (it's quite old school).
Don't hesitate if you encounter any trouble.

**Comment by pkofod** [comment]
Thank you for your reply. It was indeed the JuliaOptimizers version of CUTEst I installed, I should have made that clear. I only used the blog-post to install the homebrew-stuff. Thanks for the other pointers!

**Comment by abelsiqueira** [comment]
Hi, thanks for using my post, but it isn't updated to the last version of the packages.
One thing that is almost released is the bindeps version, that automatically install CUTEst for linux and OSX.
We're working on some changes using [NLPModels.jl](https://github.com/JuliaOptimizers/NLPModels.jl), and I've been working on it depth-first, so I haven't touched CUTEst.jl in two months. Contact me if there is something you want to expedite.

Also, I'm interested in the load problems you had.

**Comment by pkofod** [comment]
A bindeps version would be wonderful, as the whole homebrew thing was a bit, well, annoying :)

I would love to see what the following does on your installation: https://gist.github.com/pkofod/5dda61027b6829e91c40dccbe6f38f9b . On my computer, all of those problems throw an error.

**Comment by dpo** [comment]
> A bindeps version would be wonderful, as the whole homebrew thing was a bit, well, annoying :)

Did something go wrong?

> I would love to see what the following does on your installation: https://gist.github.com/pkofod/5dda61027b6829e91c40dccbe6f38f9b . On my computer, all of those problems throw an error, so works ends up being a vector of false.

What is the error? Did you try from the command line?

**Comment by abelsiqueira** [comment]
> A bindeps version would be wonderful,

It's started [here](https://github.com/abelsiqueira/CUTEst.jl/tree/feat/bindeps). Haven't tested it in a while, but it was working.

> as the whole homebrew thing was a bit, well, annoying :)

If you're on linux, I agree with you. Notice how bindeps uses [linux-cutest](https://github.com/abelsiqueira/linux-cutest), which is a simpler alternative, although very underdeveloped (no tests, no reinstall, etc.)

**Comment by abelsiqueira** [comment]
> I would love to see what the following does on your installation: https://gist.github.com/pkofod/5dda61027b6829e91c40dccbe6f38f9b . On my computer, all of those problems throw an error, so works ends up being a vector of false.

Tried it, and they all throw an error, but that's because those problems are not on `$MASTSIF`. Where did you get them?
To get all unconstrained problems, you can use
````
grep " .U" $MASTSIF/CLASSF.DB | cut -f1 -d" "
````

**Comment by pkofod** [comment]
> Did something go wrong?

No, nothing went wrong, but as a linux user, homebrew is not really something I've ever used for anything before. It all worked out (after I found the blog post), but it would of course be preferable if  `Pkg.add("CUTEst")` was all you needed :)



> Tried it, and they all throw an error, but that's because those problems are not on $MASTSIF. Where did you get them?
> To get all unconstrained problems, you can use

Okay, my bad then! The error was probably just that I didn't have them. I thought MASTSIF had all the CUTEst problems.Thanks for the grep.

**Comment by pkofod** [comment]
So, the revised test set is 173 unconstrained problems with the following dimensions:
```
nvar   count
2      28
3      14
4      4
5      1
6      3
7      1
8      9
10     3
11     1
12     1
15     1
50     8
63     1
99     1
100    2
200    7
500    1
1000   4
1024   2
2000   1
2550   2
2600   2
2652   1
3000   16
3549   1
4000   2
4900   1
4999   1
5000   36
5010   1
5625   2
10000  11
20000  2
100000 2
```
Quite a few big problems, and maybe a bit too few in the range 10-50 (I suspect a lot of users might be interested in these. I'm running stuff, but it takes a while, that's why there's been no update :)

**Comment by pkofod** [comment]
Just starting to get this under control. Ran BFGS with the 4 different line search algorithms we have. Below you see performance profiles for test problems up to (and including) dimension 5000. Seems like HZ (the default) is a pretty solid choice!
![newplot](https://cloud.githubusercontent.com/assets/8431156/16535925/0e356fe8-3fec-11e6-84dc-dcf2ca2c52da.png)

**Comment by dpo** [comment]
What are you measuring in this profile? Number of objective evals?

Ps: https://github.com/JuliaSmoothOptimizers/Profiles.jl

**Comment by pkofod** [comment]
No, this is simply for the attained minimum. Cool that you have a package for that in JSO. I just wrote it myself, as there was no license in the Matlab file on More and Wild's website. Here they are for `f_calls`
![newplot 1](https://cloud.githubusercontent.com/assets/8431156/16538858/e49f05c8-4030-11e6-8594-d6b2115a4947.png)

**Comment by dpo** [comment]
That's instructive. The two plots together tell a more complete story. Armijo remains amazingly robust given how simple it is. I'll have to look at the details of your interpolating linesearch.

**Comment by pkofod** [comment]
I would still say hz is dominating if we take into consideration the actual minimum they end up reporting.

**Comment by pkofod** [comment]
I should compare these with LBFGS. thanks again for the CUTEst package :) the newest version is even easier to install !

**Comment by pkofod** [comment]
![newplot 2](https://cloud.githubusercontent.com/assets/8431156/16541359/e67b937a-4080-11e6-87d3-ef11bb76e6f8.png)

This is a surprise... HZ linesearch isn't the best for HZ CG :)

**Comment by pkofod** [comment]
I simply can't run the largest problems on my laptop, and even the medium-large sized ones for Nelder-Mead, so I am waiting for the sysadmin to fix a symlink-thing on the server at the econ department. However, I'm starting to gather some evidence at least.
![newplot 3](https://cloud.githubusercontent.com/assets/8431156/16541867/0595402e-4092-11e6-9dde-653c3f5045ae.png)

**Comment by pkofod** [comment]
I'm just spamming at this point. Any participants are welcome to unsubscribe if they're annoyed :)

![newplot 4](https://cloud.githubusercontent.com/assets/8431156/16542384/004dfe26-40a6-11e6-94c9-2728166af7fd.png)

This is actually quite surprising to me! (I took the upper tau to 20 for clarity, the picture remains the same with other taus bars). Conjugate Gradient is doing remarkably bad compared to Gradient descent here! I've cherry-picked the best line search algorithms for each "solver" or overall algorithm.

**Comment by Evizero** [comment]
I didn't really follow along here, but just scrolling over the images made me curious. Are these for the Julia implementation in this package? if so what did you use for plotting?

**Comment by pkofod** [comment]
Yes they are. Plots.jl with the Plotly backend

**Comment by timholy** [comment]
Almost surely a bug in CG, perhaps in the convergence criteria? You already noticed one typo, there may be others!

This is what's so wonderful about having a good test suite! :100:

**Comment by pkofod** [comment]
Now CG is actually in the plot, and doing much better than gradient descent which has 4 entries in the bottom 7 - auch :) BFGS still seems to dominate, and the BFGS+HZ solver seems to be a safe default (as it is).

**Comment by mlubin** [comment]
@pkofod, where is the code you're using to run these experiments?

**Comment by pkofod** [comment]
> @pkofod, where is the code you're using to run these experiments?
> EDIT: not sure if you mean that you'll publish the performance profile code or all of it together shortly

I just mean that I'll post the code I used to produce these figures. It's only available if you're setting next to me right now. :) I'll post a gist or something. I just need to organize it a bit better.

**Comment by pkofod** [comment]
As an example, this code produces the figures posted in #220 , that I repeat here:
![newplot 10](https://cloud.githubusercontent.com/assets/8431156/16568284/81ee1b40-4228-11e6-92e3-e77f9653c059.png)

Do produce that figure, you must first load the module, Plots, and Optim. Then I construct a vector of names, and a vector of `Optimizer` specifications to be benchmarked
```julia
using Plots, OptimBenchmarks, Optim

names = ["NelderMead with fixed parameters and unit step",
         "NelderMead with fixed parameters and relative_step (0.05)",
         "NelderMead with fixed parameters and relative step (0.1)",
         "NelderMead with fixed parameters and relative step (0.50)",
         "NelderMead with adaptive parameters and relative step (0.05)",
         "NelderMead with adaptive parameters and relative step (0.1)",
         "NelderMead with adaptive parameters and relative step (0.50)"]

opt_cases =[Optim.NelderMead(parameters = Optim.fixed_parameters,    initial_step = Optim.unit_step),
            Optim.NelderMead(parameters = Optim.fixed_parameters,    initial_step = x->Optim.relative_step(x; s = 0.05)),
            Optim.NelderMead(parameters = Optim.fixed_parameters,    initial_step = x->Optim.relative_step(x; s = 0.1)),
            Optim.NelderMead(parameters = Optim.fixed_parameters,    initial_step = x->Optim.relative_step(x; s = 0.5)),
            Optim.NelderMead(parameters = Optim.adaptive_parameters, initial_step = x->Optim.relative_step(x; s = 0.05)),
            Optim.NelderMead(parameters = Optim.adaptive_parameters, initial_step = x->Optim.relative_step(x; s = 0.1)),
            Optim.NelderMead(parameters = Optim.adaptive_parameters, initial_step = x->Optim.relative_step(x; s = 0.5))]
```
then I run the benchmarks. It uses try/catches, as an out of memory error or line search error would otherwise ruin everything. We should make line searches soft quit imo.
```julia
out = benchmark(names, opt_cases, Optim.OptimizationOptions(iterations = 20000); max_dim = 1000)

save("nelder_mead_benchmark.csv", out...)
```
Then, load the data as a dataframe (I just saved it to keep a copy), and plot it
```julia
dff = read_profile_data("nelder_mead_benchmark.csv")

tau = 1:0.5:20
profiles = performance_profile(names, dff, tau)

plot(tau, hcat(profiles...),
          label = hcat(names...),
          ls=[:solid  :dash :dash :dash :dot :dot :dot],
          size =(800,500),
          ylims = (0,1))
```

the module is in https://gist.github.com/pkofod/82ba6021e73cf9ee2289f35b08a09f2e (some things could be done more cleanly...)

**Comment by pkofod** [comment]
This work is now in OptimTests, thanks for the feedback.
